### PR TITLE
SI-7008 @throws annotations are now populated in reflect

### PR DESCRIPTION
**Pull request by xeno-by**
Runtime reflection in JavaMirrors previously forgot to fill in
@throws when importing Java reflection artifacts. Now this is fixed.

Note that generic exception types used in `throws` specifications will
be garbled (i.e. erased), because we don't use `getGenericExceptionTypes`
in favor of just `getExceptionTypes` to stay compatible with the behavior
of ClassfileParser. That's a bug, but a separate one and should be fixed
separately.

Also note that this commit updated javac-artifacts.jar, because we need
to test how reflection works with javac-produced classfiles. The sources
that were used to produce those classfiles can be found in the jar next
to the classfiles.

### CODE REVIEW

**Comment by xeno-by** [comment]
review @gkossakowski @retronym @paulp

**Comment by gkossakowski** [comment]
@xeno-by regarding `javac-artifacts.jar`, is there any reason you cannot follow the setup I used in #2021 where use `_1`, `_2`, etc. file suffixes to compile Java files first and use them on classpath for the test?

**Comment by xeno-by** [comment]
@gkossakowski That's a good question. Back then I had problems with that, but I cannot remember their nature. Are `java` files in partests compiled by `javac` or by `scalac`?

**Comment by gkossakowski** [comment]
@xeno-by: they are compiled with javac. Scalac can only parse Java files to populate symbol table for mixed compilation. It doesn't compile Java files :)

It would be useful to try. I prefer everything to be in source code. Eventually, we should get rid of `javac-artifacts.jar` altogether.

**Comment by xeno-by** [comment]
I agree. If there are no problems, I don't see a reason why we should be roundabout with all those auxiliary jars.

**Comment by xeno-by** [comment]
@gkossakowski Btw I noticed that your nearby pull request changes how `ClassfileParser` creates `@throws` annotations. This PR should do exactly the same. What do we do about this dependency?

**Comment by gkossakowski** [comment]
@xeno-by: yes, I agree it should exactly the same. What problem do you see here? Or do you want to check if Scala runtime reflection produces the same result as Scala compile time reflection?

**Comment by xeno-by** [comment]
I just wanted to know whether you're done with that PR, so that I can rebase off it and reuse the changed logic of `ClassfileParser` in `JavaMirrors`.

**Comment by gkossakowski** [comment]
@xeno-by: Yes, I think it should be merged as-is. I'll ping @adriaanm to look at it ASAP.

**Comment by xeno-by** [comment]
@gkossakowski Thanks!

**Comment by xeno-by** [comment]
@gkossakowski @adriaanm There's a problem, though. Unfortunately, I cannot do `typer.packSymbols(cls.typeParams, cls.tpe)` in runtime reflection. What would you advise to do instead?

**Comment by paulp** [comment]
I propose:

```

scala> import scala.reflect.runtime.universe._
import scala.reflect.runtime.universe._

scala> val clazz = typeOf[scala.throws[_]].typeSymbol.asClass
clazz: reflect.runtime.universe.ClassSymbol = class throws

scala> existentialAbstraction(clazz.typeParams, clazz.toType)
res0: reflect.runtime.universe.Type = throws[_ <: Throwable]
```

**Comment by xeno-by** [comment]
@paulp done

**Comment by gkossakowski** [comment]
@paulp: Interesting. Should I use the same thing in `ClassfileParser` in #2021 ?

**Comment by xeno-by** [comment]
@gkossakowski I already did that - the logic is now shared between `JavaMirrors` and `ClassfileParser`.

**Comment by gkossakowski** [comment]
@xeno-by: I didn't notice. Thanks!

BTW. I see that you still refer to java types from `javac-artifacts.jar`. The idea to compile everything from source didn't fly?

**Comment by xeno-by** [comment]
Didn't have time for that yet. Will do that today or tomorrow.

**Comment by xeno-by** [comment]
Done. That was easier than I thought :)

**Comment by gkossakowski** [comment]
@xeno-by legitimate failure.

File: src/reflect/scala/reflect/internal/AnnotationInfos.scala [hunk 1]
```diff
@@ -33,6 +33,17 @@ trait AnnotationInfos extends api.Annotations { self: SymbolTable =>
       case ThrownException(exc) => exc
     }
 
+    def addThrowsAnnotation(excSym: Symbol): Self = {
```
Comment (retronym) [inline src/reflect/scala/reflect/internal/AnnotationInfos.scala]: Can we pick something more expressive for `excSym` / `excTpe`? e.g. `throwableSym/Tpe`?

File: src/reflect/scala/reflect/runtime/JavaMirrors.scala [hunk 2]
```diff
@@ -610,11 +610,19 @@ private[reflect] trait JavaMirrors extends internal.SymbolTable with api.JavaUni
 
     /**
      * Copy all annotations of Java annotated element `jann` over to Scala symbol `sym`.
+     * Also creates `@throws` annotations if necessary.
      *  Pre: `sym` is already initialized with a concrete type.
      *  Note: If `sym` is a method or constructor, its parameter annotations are copied as well.
      */
     private def copyAnnotations(sym: Symbol, jann: AnnotatedElement) {
       sym setAnnotations (jann.getAnnotations map JavaAnnotationProxy).toList
+      // FIXME: we're not using getGenericExceptionTypes here to be consistent with ClassfileParser
```
Comment (retronym) [inline src/reflect/scala/reflect/runtime/JavaMirrors.scala]: When would this be fixed? Let's track it in a ticket.

File: src/reflect/scala/reflect/runtime/JavaMirrors.scala [hunk 3]
```diff
@@ -610,11 +610,19 @@ private[reflect] trait JavaMirrors extends internal.SymbolTable with api.JavaUni
...
+      val jexTpes = jann match {
+        case jm: jMethod => jm.getExceptionTypes.toList
+        case jconstr: jConstructor[_] => jconstr.getExceptionTypes.toList
+        case _ => Nil
+      }
+      jexTpes foreach (jexTpe => sym.addThrowsAnnotation(classSymbol(jexTpe)))
```
Comment (retronym) [inline src/reflect/scala/reflect/runtime/JavaMirrors.scala:625]: How are Scala classes containing `@throws` annotations treated? I can't figure out whether we pickle the annotation in addition to adding the exception to the signature. If we do, might we end up with duplicate annotations in runtime reflection? This warrants a test.

File: src/reflect/scala/reflect/runtime/JavaMirrors.scala (hunk 2 above)
Comment (xeno-by) [inline src/reflect/scala/reflect/runtime/JavaMirrors.scala]: Done

File: src/reflect/scala/reflect/runtime/JavaMirrors.scala (hunk 3 above)
Comment (xeno-by) [inline src/reflect/scala/reflect/runtime/JavaMirrors.scala:625]: Scala classes are unpickled with a different completer.

File: src/reflect/scala/reflect/internal/AnnotationInfos.scala (hunk 1 above)
Comment (xeno-by) [inline src/reflect/scala/reflect/internal/AnnotationInfos.scala]: Done

File: src/reflect/scala/reflect/runtime/JavaMirrors.scala (hunk 3 above)
Comment (xeno-by) [inline src/reflect/scala/reflect/runtime/JavaMirrors.scala:625]: I've added a test anyways :)

**Comment by xeno-by** [comment]
All done, Jason. Can I get my rubber stamp now? :)

**Comment by retronym** [comment]
LGTM :ok: :+1:

**Comment by xeno-by** [comment]
@adriaanm Why does the UI say "Good to merge", while there's pr-checkin-per-commit in flight?

**Comment by adriaanm** [comment]
Because github only looks at the most recent status in a commit.
I'm working on reconfiguring jenkins so that we only have one job per commit now.

**Pull request merged**